plots: document html template (#2472)
iterative/dvc.org: content/docs/user-guide/plotting-in-offline-environments.md

# Plotting in offline environments

There may be cases when the user does not have an Internet connection for
various reasons. `dvc plots` - generated HTML will not render properly in these
cases. This guide shows how to fix plotting in offline environments.

DVC will generate HTML file that contains metrics visualizations when using
[plots](/doc/commands-reference/plots). The file contains a JavaScript code that
employs [Vega-Lite](https://vega.github.io/vega-lite/) to present the plot. The
required JavaScript libraries for this HTML file are downloaded from the web:

```html
<script src="https://cdn.jsdelivr.net/npm/vega@5.10.0"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-lite@4.8.1"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-embed@6.5.1"></script>
```

DVC allows to provide a custom HTML template to avoid this requirement.

> ## Example: Setup local server with Vega libraries

Author: This was my initial part, but I like the server version more due to its robustness:
If we specify server, we are pretty sure the address won't change for the same network. In case of local path, we need to assure two things:

1. that template path is absolute path (`dvc plots` produces html's in cwd, so relative paths would require some special logic dedicated to resolving vega libs paths according to `cwd`) 
2. that all users copying given repo will have the vega libs under the same path

Using local server, I intend to point user to a solution that would be viable for intranet and would work without every user having to put libs in same place, or modify the project each time they change the machine they work on.

> ## Example: Setup local server with Vega libraries

Reviewer: The final plots document will always need the webserver running in this case, right? 

For the intranets, I see your points. That's surely more robust but that configuration can be completely different, e.g., they may already have a central URL to put the files and we don't expect users to copy-paste this template to their workflow. 

We just want to show how to supply a template and the user might be looking for a way to render their plots completely differently. 

I don't have a strong opinion on this and will trust your judgment. If you see it's better this way, let's keep it. For me, although it has some extra parts, the guide satisfies the purpose of describing the templates. Thank you.

> ## Example: Setup local server with Vega libraries

Author: I agree that server approach is overkill for example purposes, and tries to direct user into "proper" direction. Maybe lets make an example for local path, and just add note on the bottom of the page why one should consider "static" URL to js?

> ## Example: Setup local server with Vega libraries

Reviewer: If you have time for this, could you edit the document to move the server part as an additional section? I can take a look at it on Monday or Tuesday as well.

> ## Example: Setup local server with Vega libraries

Author: I will try to fix that on monday

> ## Example: Setup local server with Vega libraries

Author: Hey, I managed to modify the guide to use local files.

> ## Example: Setup local server with Vega libraries

Reviewer 2: @iesahin I'm still seeing the web server section so not sure what the conclusion was here.

# Plotting in offline environments

> There may be cases when the user does not have an Internet connection for
> various reasons. `dvc plots` - generated HTML will not render properly in these
> cases. This guide shows how to fix plotting in offline environments.

Reviewer 2: TBH I didn't review the new guide until now since I thought the plan was to extract it from this PR. In general it needs some syntax check, simplification, summarizing, linking, etc., I think e.g.

```suggestion
If there' no Internet connection, the HTML file generated by `dvc plots` will not render properly.
This guide shows how to display plots when offline.
```

Reviewer 2: TBH I also wouldn't introduce the new guide yet (there's #2572 to figure that out) but if others approve, then OK but let's just rewrite it per the guidelines above based on this text that @pared provided? Thanks

Reviewer: [Here](https://github.com/iterative/dvc.org/pull/2472#issuecomment-865811690) I said

> I'll delete the UG file if everything looks all right. 

I don't intend to keep that guide.

Reviewer 2: > I'll delete the UG file if everything looks all right.

Ah OK cool please do then 🙂
